### ISSUE TITLE
The trans Performance Problem in Certain Scenarios

### ISSUE THREAD

**Issue by h1074112368** [triton-lang/triton-ascend#1611]
### Describe the issue

```python
#!/usr/bin/env python3
"""Focused head-to-head: 13*16 vs 13*12 (3D bare load+trans+store form).

Self-contained (no imports from other workspace scripts). Kernel: block_ptr
[1, RB, XB] + tl.trans(t, (2, 1, 0)) + transposed store, RB=16 (RB=64
overflows UB in the 3D form), SQ=SK=256; bit-exact permutation
out[sq, h, sk] = b[sq, sk, h].

  13*16: single trip (XB=16 > H=13, read width clamped to 13, 52B pitch dense)
  13*12: H % XB != 0 -> tail trip at x_off=12 with ONE valid column still
         fetching full rows

Measured: 13*16 = 748.74us [0.2%], 13*12 = 926.09us [0.7%] ->
13*12 / 13*16 = 1.24x, consistent with the reduction form (1.24x) and the
multi-group run (1.27x).

Validation:
  Precision: dual reference -- fp32 eager (bit-exact permutation -> err32
    must be 0, hard assert) + fp64 CPU truth (the NPU has no double, so the
    ground truth is computed on the CPU; a pure fp32 copy is exact in fp64
    too -> err64 must be 0, hard assert).
  Performance: 5 rounds x 50 iters per case, med[min, max] + spread + GB/s
    reported; 13*12 must be <= 1.05x of the 13*16 baseline (hard assert,
    FAIL -> exit 1).

Usage: source env.sh keep && python _trans_align_pair_en.py
"""
import importlib.util
import os

import torch
import torch_npu  # noqa: F401

SQ, SK = 256, 256     # aligned with profile_t5 small (B=2, S=256): pos_bias
                      # has Sq=256 rows (no batch dim), Sk=256
RB = 16               # 3D form: RB=64 overflows UB (4194304 > 1572864 bits at
                      # 64*64); RB=16 keeps the tile inside the 192KB UB budget
KDIR = "_bench_kernels"
ITERS, ROUNDS = 50, 5

KERNEL = '''
import triton
import triton.language as tl

@triton.jit
def trans_load_store(b_ptr, out_ptr, H: tl.constexpr, SK: tl.constexpr,
                     RB: tl.constexpr, XB: tl.constexpr):
    """Bare load + trans + store: out[sq, h, sk] = b[sq, sk, h].

    No reduction, no accumulation -- only the 3D block_ptr load, the UB
    permutation tl.trans(t, (2, 1, 0)) and the transposed store. Row tile
    fixed at 1 (the real kernel tiles it, here it is the program dim).
    Read width XB_READ = min(H, ceil32B(XB)) keeps rows dense; the int
    axis is chunked by the logical XB -- H % XB != 0 produces the tail
    trip whose tile has few valid columns but still fetches whole rows.
    """
    XBR: tl.constexpr = ((XB + 7) // 8) * 8      # round up to a 32B multiple
    XB_READ: tl.constexpr = XBR if XBR < H else H
    pid = tl.program_id(0)          # sq row
    base = b_ptr + pid * (SK * H)   # this row's b segment: contiguous [SK, H] block
    for x_off in tl.range(0, H, XB):
        for r_off in tl.range(0, SK, RB):
            bp = tl.make_block_ptr(base, shape=[1, SK, H], strides=[H * SK, H, 1],
                                   offsets=[0, r_off, x_off],
                                   block_shape=[1, RB, XB_READ], order=[2, 1, 0])
            t = tl.load(bp, boundary_check=[0, 1, 2], padding_option="zero")
            g = tl.trans(t, (2, 1, 0))   # [1, RB, XB_READ] -> [XB_READ, RB, 1]
            # store directly in 3D (no reshape staging -- the reshape pushed
            # the 64*64 tile over the UB budget)
            tl.store(out_ptr + pid * (H * SK)
                     + (x_off + tl.arange(0, XB_READ))[:, None, None] * SK
                     + (r_off + tl.arange(0, RB))[None, :, None],
                     g,
                     mask=(x_off + tl.arange(0, XB_READ) < H)[:, None, None])
'''


def make_kernel(h, xb):
    src = KERNEL
    os.makedirs(KDIR, exist_ok=True)
    path = os.path.join(KDIR, f"trans_bare3d_h{h}_xb{xb}.py")
    with open(path, "w") as f:
        f.write(src)
    spec = importlib.util.spec_from_file_location(f"trans_bare3d_h{h}_xb{xb}", path)
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    return mod.trans_load_store


def bench_rounds(kern, grid, args, kwargs, iters=ITERS, rounds=ROUNDS):
    """rounds independent Event timings (ms/iter), returns (median, min, max)."""
    torch.npu.synchronize()
    for _ in range(3):
        kern[grid](*args, **kwargs)
    torch.npu.synchronize()
    times = []
    for _ in range(rounds):
        t0 = torch.npu.Event(enable_timing=True)
        t1 = torch.npu.Event(enable_timing=True)
        t0.record()
        for _ in range(iters):
            kern[grid](*args, **kwargs)
        t1.record()
        torch.npu.synchronize()
        times.append(t0.elapsed_time(t1) / iters)
    times.sort()
    return times[len(times) // 2], times[0], times[-1]


def run_bare(shape):
    """Compile (cached), validate the exact permutation, bench; returns med ms."""
    h, xb = shape
    torch.manual_seed(0)
    b = torch.randn(SQ, SK, h, device="npu")
    out = torch.empty(SQ, h, SK, device="npu")
    ref32 = b.permute(0, 2, 1)          # [SQ, H, SK] -- exact copy, bit-identical
    # fp64 ground truth on the CPU: the NPU has no double dtype (it silently
    # casts to float, which would make err64 always equal err32 and void the
    # check -- see the ToKernelNpu.cpp warning). A pure fp32 copy is exact in
    # fp64 too, so err64 must be 0 as well.
    ref64 = b.cpu().double().permute(0, 2, 1)
    kern = make_kernel(h, xb)
    kern[(SQ,)](b, out, H=h, SK=SK, RB=RB, XB=xb)
    torch.npu.synchronize()
    # Pure permutation: the result must be bit-exact against BOTH references.
    err32 = (out - ref32).abs().max().item()
    err64 = (out.cpu().double() - ref64).abs().max().item()
    assert err32 == 0.0, f"H={h} XB={xb}: permutation err32 {err32:.2e} != 0"
    assert err64 == 0.0, f"H={h} XB={xb}: permutation err64 {err64:.2e} != 0"
    med, lo, hi = bench_rounds(kern, (SQ,), (b, out),
                               {"H": h, "SK": SK, "RB": RB, "XB": xb})
    spread = (hi - lo) / med * 100
    rbytes = SQ * SK * h * 4
    wbytes = SQ * h * SK * 4
    gbs = (rbytes + wbytes) / (med * 1e-3) / 1e9
    print(f"shape {h:>2}x{xb:<2}  med {med*1000:7.2f} us  [{lo*1000:7.2f}, {hi*1000:7.2f}]  "
          f"{spread:4.1f}%  {gbs:5.1f} GB/s  err32 {err32:.1e}  err64 {err64:.1e}", flush=True)
    return med


def main():
    # case 1: 13*16 -- single trip, 52B pitch dense
    # device time ~749 us (0.749 ms, measured)
    time_16 = run_bare((13, 16))

    # case 2: 13*12 -- tail trip: 1 valid column, full-row fetch
    # device time ~951 us (0.951 ms, measured)
    time_12 = run_bare((13, 12))

    print("-" * 80)
    ratio = time_12 / time_16
    print(f"13*16: {time_16*1000:.2f} us")
    print(f"13*12: {time_12*1000:.2f} us")
    print(f"13*12 / 13*16 device-time ratio = {ratio:.2f}x")
    assert time_12 <= time_16 * 1.05, (
        f"13*12 ({time_12*1000:.2f} us) exceeds 1.05x of the 13*16 baseline "
        f"({time_16*1000:.2f} us): {ratio:.2f}x -- the tail trip re-reads full "
        f"rows for one valid column")
    print("PASS: 13*12 within 5% of the 13*16 baseline")
    print("DONE")


if __name__ == "__main__":
    main()
```

### Environment details

conda env        : conda_inductor_2.13
python           : 3.11.15
torch            : 2.13.0+cpu
torch_npu        : 2.13.0+git9e5a97a
triton-ascend    : 3.2.2+git0d69f740   (whl: triton_ascend-3.2.2+git0d69f740)
CANN             : 9.1.0-beta.1
bishengir-compile: 1.1.0 (AscendNPU-IR, commit 7058cef3d997, LLVM 19.1.7)
driver           : 26.0.rc1 (ascendhal 7.35.23)
device           : Ascend910B2 x 8, 64GB HBM/卡, Cube 24 核 + Vector 48 核, L2 192MB
OS               : Ubuntu 22.04.5, kernel 6.6.0-132, aarch64

**Comment by hongziqi** [comment]
Thanks for your feedback. We will arrange relevant engineers to investigate this issue later.

**Comment by strategist614** [comment]
## Investigation update

I did several controlled experiments on the `H=13, XB=12` regression and the current evidence suggests that the main issue is **overlapping iteration coverage caused by a mismatch between the loop step (`XB`) and the actual tile width (`XB_READ`)**, rather than a full-row tail reload in `LoadStoreConverter`.

### Original kernel behavior

For the repro:

```python
H = 13
XB = 12

XBR = ((XB + 7) // 8) * 8   # 16
XB_READ = min(XBR, H)        # 13
```

the outer loop is:

```python
for x_off in tl.range(0, H, XB):
```

so the loop executes with:

```text
x_off = 0
x_off = 12
```

However, each iteration uses a block width of `XB_READ = 13`.

Therefore:

```text
x_off = 0
  covers columns [0, 13)
  -> columns 0..12 are already fully processed

x_off = 12
  nominally covers [12, 25)
  -> after boundary handling, only column 12 is valid
  -> column 12 is processed again
```

So the second iteration is semantically redundant for this case.

### Performance

Original repro:

```text
13x16: ~359 us
13x12: ~445 us
ratio: ~1.24x
```

If the redundant second `x` iteration is removed, while keeping correctness checks enabled:

```text
13x16: 358.11 us
13x12: 357.81 us
ratio: 1.00x

err32 = 0
err64 = 0
```

This is the strongest result so far: removing only the redundant `x=12` iteration completely eliminates the regression without changing the output.

### Store-side transpose experiment

Initially I suspected that the tail store was expensive because the generated IR performed:

```text
full transpose
    ->
extract_slice(valid tail)
```

I changed the masked store lowering to push the slice through `linalg.transpose`, producing:

```text
extract_slice(input)
    ->
transpose(smaller tensor)
```

For the tail iteration the IR successfully changed from a full-width transpose to:

```mlir
%extracted_slice =
  tensor.extract_slice %28[0, 0, 0] [1, 16, %33] [1, 1, 1]
    : tensor<1x16x13xf32> to tensor<1x16x?xf32>

%34 = tensor.empty(%33) : tensor<?x16x1xf32>

%transposed =
  linalg.transpose
    ins(%extracted_slice : tensor<1x16x?xf32>)
    outs(%34 : tensor<?x16x1xf32>)
    permutation = [2, 1, 0]
```

However performance remained approximately:

```text
13x16: ~359 us
13x12: ~447 us
```

So the full transpose was not the main cause.

### Padding-fill experiment

The boundary-load lowering allocates a static temporary buffer and conditionally fills the full buffer for padding before copying the valid subview.

I temporarily removed the full-tile padding fill as a diagnostic experiment.

Performance remained:

```text
13x16: ~359 us
13x12: ~445 us
```

So the padding fill is also not the main source of the regression.

### `tl.static_range` experiment

I also replaced:

```python
for x_off in tl.range(0, H, XB):
```

with:

```python
for x_off in tl.static_range(0, H, XB):
```

This fully unrolled the outer `x` loop.

The generated IR contained two independent inner loops:

```text
x = 0:
  scf.for r = 0 to 256 step 16

x = 12:
  scf.for r = 0 to 256 step 16
```

The second tail body was completely static:

```mlir
tensor.extract_slice ... [1, 16, 1]
    : tensor<1x16x13xf32> to tensor<1x16x1xf32>

tensor.empty() : tensor<1x16x1xf32>

linalg.transpose
    tensor<1x16x1xf32>
      -> tensor<1x16x1xf32>

memref.subview ... [1, 16, 1]
```

There were no dynamic tensor dimensions in the tail body.

Nevertheless:

```text
13x16: 358.83 us
13x12: 442.54 us
ratio: 1.23x
```

Therefore the regression is not primarily caused by the surviving dynamic outer `scf.for` or dynamic tail shapes.

### Current interpretation

The important observation is that `XB=12` is used as the iteration step while the actual tile width is expanded to `XB_READ=13`.

That creates overlapping coverage:

```text
iteration 0:  [0 ............ 12]
iteration 1:                  [12]
                               ^
                         redundant work
```

The second `x=12` iteration also executes the full inner loop:

```text
SK = 256
RB = 16

256 / 16 = 16 inner iterations
```

Therefore one redundant output column causes another 16 inner-loop bodies to execute.

This explains the approximately 24% regression much better than the original hypothesis that the tail iteration re-reads a full row.

### Possible fix / question

At the kernel level, changing the outer step to match the actual tile coverage removes the redundant iteration:

```python
for x_off in tl.range(0, H, XB_READ):
```

For this repro this produces one iteration and restores the `13x12` runtime to the `13x16` baseline.

The open question is whether Triton-Ascend should attempt to optimize this case in the compiler.

Doing so would require proving that the later iteration is fully redundant despite overlapping loads/stores, which seems closer to cross-iteration range/coverage analysis than a local `LoadStoreConverter` optimization.

So at this point I think there are two possible interpretations:

1. The kernel should guarantee that the loop step is at least the effective tile coverage (`XB_READ`), in which case this is primarily a kernel-side issue.

2. Triton-Ascend may want an optimization for small constant-trip loops with overlapping accesses, but that would be a more general loop/access optimization rather than a load/store lowering fix.

I would appreciate feedback on which behavior is expected here, especially whether the compiler is expected to eliminate this kind of redundant overlapping tail iteration.

**Comment by strategist614** [comment]
I did a few additional experiments and the slowdown appears to come from a redundant overlapping tail iteration rather than an Ascend-specific load/store lowering issue.

For the original case:

```python
H = 13
XB = 12
XB_READ = 13
```

but the loop uses:

```python
for x_off in tl.range(0, H, XB):
```

So `x=0` already covers columns `0..12`, while `x=12` enters another iteration and redundantly processes column `12`.

Changing the loop to:

```python
for x_off in tl.range(0, H, XB_READ):
```

removes the regression on Ascend 910B2:

```text
XB=16:        ~358 us
XB=12:        ~443-445 us
XB=12 fixed:  ~358 us
```

I also reproduced the same pattern with Triton CUDA:

```text
RTX 2080 / SM75:  +56.8%
RTX 3090 / SM86:  +11.5% to +18.2%
RTX 4090 D / SM89: +8.6% to +11.7%
Ascend 910B2:     ~+24%
```

CUDA TTIR/TTGIR also preserves the redundant `x=12` iteration, so this behavior does not appear to be specific to Triton-Ascend.

The current conclusion is:

> The slowdown is caused by overlapping iteration coverage (`step=12`, effective tile width=13), which creates a redundant tail iteration. The straightforward kernel-side fix is to make the loop step consistent with the effective tile coverage.
